File: bin/generate_bulk_fix_module.py
#!/usr/bin/env python3
"""Generates lib/medoru/content/kanji_bulk_fix.ex with embedded data from JSON."""

import json
import os

PROJECT = "/var/home/meddle/development/elixir/medoru"

# Load all fix sources
stroke_fixes = {}
reading_fixes = {}

# 1. KanjiVG stroke fixes (best quality)
with open(f"{PROJECT}/priv/repo/seeds/bulk_stroke_fixes.json", "r", encoding="utf-8") as f:
    all_kvg = json.load(f)

# 2. Direction fixes for kanji with unknown directions
with open(f"{PROJECT}/priv/repo/seeds/bulk_direction_fixes.json", "r", encoding="utf-8") as f:
    all_direction = json.load(f)

# 3. Reading fixes from kanjidic2
with open(f"{PROJECT}/priv/repo/seeds/bulk_reading_fixes.json", "r", encoding="utf-8") as f:
    all_readings = json.load(f)

# Read chars that need fixing
with open("/tmp/chars_strokes.txt", "r", encoding="utf-8") as f:
    stroke_chars = set(f.read().strip().split("\n"))

with open("/tmp/chars_readings.txt", "r", encoding="utf-8") as f:
    reading_chars = set(f.read().strip().split("\n"))

# Build stroke fixes: prefer KanjiVG for types, but use direction fixes for directions
# For KanjiVG data, if direction is unknown, infer a reasonable default from type
TYPE_TO_DIRECTION = {
    "horizontal": "left-to-right",
    "vertical": "top-to-bottom",
    "diagonal": "top-left-to-bottom-right",
    "corner": "top-left-to-bottom-right",
    "dot": "top-to-bottom",
    "hook": "top-to-bottom",
    "curve": "top-left-to-bottom-right",
    "rising": "bottom-left-to-top-right",
}

for char in stroke_chars:
    if char in all_kvg:
        # Start with KanjiVG data
        kvg_data = all_kvg[char]
        improved = []
        for s in kvg_data["strokes"]:
            d = s["direction"]
            t = s["type"]
            if d == "unknown" and t in TYPE_TO_DIRECTION:
                d = TYPE_TO_DIRECTION[t]
            improved.append({
                "path": s["path"],
                "order": s["order"],
                "type": t,
                "direction": d,
            })
        stroke_fixes[char] = {
            "strokes": improved,
            "bounds": kvg_data["bounds"],
        }
    elif char in all_direction:
        stroke_fixes[char] = all_direction[char]

# Build reading fixes
# For katakana readings, force type to :on since the validator requires katakana for on-readings
def is_katakana(s):
    return all('\u30a0' <= c <= '\u30ff' for c in s)

for char in reading_chars:
    if char in all_readings:
        readings = []
        for r in all_readings[char]:
            r_type = r["type"]
            if r_type == "kun" and is_katakana(r["reading"]):
                r_type = "on"
            readings.append({"type": r_type, "reading": r["reading"], "romaji": r["romaji"]})
        reading_fixes[char] = readings

# Add manual fixes for chars not in kanjidic2
# 犰: used in 犰狳 (armadillo), reading キュウ
if "犰" in reading_chars:
    reading_fixes["犰"] = [{"type": "on", "reading": "キュウ", "romaji": "kyuu"}]

# 涮: used in 涮涮鍋, reading セン
if "涮" in reading_chars:
    reading_fixes["涮"] = [{"type": "on", "reading": "セン", "romaji": "sen"}]

print(f"Stroke fixes: {len(stroke_fixes)} (KanjiVG: {len([c for c in stroke_fixes if c in all_kvg])}, direction: {len([c for c in stroke_fixes if c in all_direction])})")
print(f"Reading fixes: {len(reading_fixes)}")

# Build module lines
lines = []
lines.append("defmodule Medoru.Content.KanjiBulkFix do")
lines.append('  @moduledoc """')
lines.append("  Self-contained bulk fix for kanji strokes and readings.")
lines.append("  Run: Medoru.Content.KanjiBulkFix.apply()")
lines.append('  """')
lines.append("")
lines.append("  alias Medoru.Repo")
lines.append("  alias Medoru.Content.{Kanji, KanjiReading}")
lines.append("  import Ecto.Query")
lines.append("  require Logger")
lines.append("")

# Stroke fixes
lines.append("  @stroke_fixes %{")
for char in sorted(stroke_fixes.keys()):
    strokes = stroke_fixes[char]["strokes"]
    if not strokes:
        continue
    lines.append(f'    "{char}" => [')
    for s in strokes:
        path = s["path"].replace("\\", "\\\\").replace('"', '\\"')
        lines.append(
            f'      %{{path: "{path}", order: {s["order"]}, type: "{s["type"]}", direction: "{s["direction"]}"}},'
        )
    lines.append("    ],")
lines.append("  }")
lines.append("")

# Reading fixes
lines.append("  @reading_fixes %{")
for char in sorted(reading_fixes.keys()):
    readings = reading_fixes[char]
    if not readings:
        continue
    lines.append(f'    "{char}" => [')
    for r in readings:
        lines.append(
            f'      %{{type: :{r["type"]}, reading: "{r["reading"]}", romaji: "{r["romaji"]}"}},'
        )
    lines.append("    ],")
lines.append("  }")
lines.append("")

# Functions
lines.extend([
    "  def apply do",
    "    apply_stroke_fixes()",
    "    apply_reading_fixes()",
    "  end",
    "",
    "  defp apply_stroke_fixes do",
    '    Logger.info("Applying stroke fixes...")',
    "",
    "    string_stroke_kanji =",
    "      Repo.all(Kanji)",
    '      |> Enum.filter(fn k ->',
    '        strokes = get_in(k.stroke_data, ["strokes"])',
    "        is_list(strokes) and length(strokes) > 0 and is_binary(hd(strokes))",
    "      end)",
    "",
    "    {fixed_kv, fixed_fb} =",
    "      Enum.reduce(string_stroke_kanji, {0, 0}, fn kanji, {kv, fb} ->",
    "        case Map.fetch(@stroke_fixes, kanji.character) do",
    "          {:ok, strokes} ->",
    "            current = kanji.stroke_data || %{}",
    '            new_data = Map.merge(current, %{"strokes" => strokes, "bounds" => %{"width" => 109, "height" => 109, "viewBox" => "0 0 109 109"}})',
    "            kanji |> Ecto.Changeset.change(stroke_data: new_data) |> Repo.update!()",
    "            {kv + 1, fb}",
    "",
    "          :error ->",
    '            strokes = kanji.stroke_data["strokes"] || []',
    "            new_strokes = Enum.with_index(strokes, 1) |> Enum.map(fn {path, idx} ->",
    '              %{"path" => path, "order" => idx, "type" => "unknown", "direction" => "unknown"}',
    "            end)",
    "            current = kanji.stroke_data || %{}",
    '            kanji |> Ecto.Changeset.change(stroke_data: Map.put(current, "strokes", new_strokes)) |> Repo.update!()',
    "            {kv, fb + 1}",
    "        end",
    "      end)",
    "",
    "    unknown_stroke_kanji =",
    "      Repo.all(Kanji)",
    '      |> Enum.filter(fn k ->',
    '        strokes = get_in(k.stroke_data, ["strokes"])',
    '        is_list(strokes) and is_map(hd(strokes)) and (hd(strokes)["type"] == "unknown" or hd(strokes)["direction"] == "unknown")',
    "      end)",
    "      |> Enum.filter(fn k -> Map.has_key?(@stroke_fixes, k.character) end)",
    "",
    "    replaced = Enum.reduce(unknown_stroke_kanji, 0, fn kanji, count ->",
    "      strokes = @stroke_fixes[kanji.character]",
    "      current = kanji.stroke_data || %{}",
    '      new_data = Map.merge(current, %{"strokes" => strokes, "bounds" => %{"width" => 109, "height" => 109, "viewBox" => "0 0 109 109"}})',
    "      kanji |> Ecto.Changeset.change(stroke_data: new_data) |> Repo.update!()",
    "      count + 1",
    "    end)",
    "",
    "    no_stroke_kanji =",
    "      Repo.all(Kanji)",
    '      |> Enum.filter(fn k ->',
    '        is_nil(k.stroke_data) or k.stroke_data == %{} or is_nil(k.stroke_data["strokes"]) or k.stroke_data["strokes"] == []',
    "      end)",
    "      |> Enum.filter(fn k -> Map.has_key?(@stroke_fixes, k.character) end)",
    "",
    "    added = Enum.reduce(no_stroke_kanji, 0, fn kanji, count ->",
    "      strokes = @stroke_fixes[kanji.character]",
    '      new_data = %{"strokes" => strokes, "bounds" => %{"width" => 109, "height" => 109, "viewBox" => "0 0 109 109"}}',
    "      kanji |> Ecto.Changeset.change(stroke_data: new_data) |> Repo.update!()",
    "      count + 1",
    "    end)",
    "",
    '    Logger.info("Strokes: #{fixed_kv} KanjiVG, #{fixed_fb} fallback, #{replaced} replaced, #{added} added")',
    "  end",
    "",
    "  defp apply_reading_fixes do",
    '    Logger.info("Applying reading fixes...")',
    "",
    "    kanji_without_readings =",
    "      from(k in Kanji,",
    "        left_join: r in assoc(k, :kanji_readings),",
    '        where: not is_nil(k.stroke_data) and k.stroke_data != ^%{},',
    "        group_by: k.id,",
    "        having: count(r.id) == 0,",
    "        select: k",
    "      )",
    "      |> Repo.all()",
    "",
    "    inserted =",
    "      Enum.reduce(kanji_without_readings, 0, fn kanji, count ->",
    "        case Map.fetch(@reading_fixes, kanji.character) do",
    "          {:ok, readings} when is_list(readings) and readings != [] ->",
    "            n = Enum.reduce(readings, 0, fn r, inner ->",
    '              attrs = %{kanji_id: kanji.id, reading_type: r.type, reading: r.reading, romaji: r.romaji}',
    "              case %KanjiReading{} |> KanjiReading.changeset(attrs) |> Repo.insert() do",
    "                {:ok, _} -> inner + 1",
    "                {:error, _} -> inner",
    "              end",
    "            end)",
    "            count + n",
    "",
    "          _ -> count",
    "        end",
    "      end)",
    "",
    '    Logger.info("Inserted #{inserted} readings")',
    "  end",
    "end",
])

output = "\n".join(lines) + "\n"
out_path = f"{PROJECT}/lib/medoru/content/kanji_bulk_fix.ex"
with open(out_path, "w", encoding="utf-8") as f:
    f.write(output)

size_kb = len(output) / 1024
print(f"\nWritten: {out_path}")
print(f"Size: {size_kb:.1f} KB")
print(f"Kanji with stroke fixes embedded: {len(stroke_fixes)}")
print(f"Kanji with reading fixes embedded: {len(reading_fixes)}")

File: bin/generate_bulk_fix_module.exs
#!/usr/bin/env elixir
# Generates lib/medoru/content/kanji_bulk_fix.ex with embedded data
# Usage: mix run bin/generate_bulk_fix_module.exs

alias Medoru.Repo
alias Medoru.Content.Kanji
import Ecto.Query

stroke_fixes_file = "priv/repo/seeds/bulk_stroke_fixes.json"
reading_fixes_file = "priv/repo/seeds/bulk_reading_fixes.json"

stroke_fixes =
  case File.read(stroke_fixes_file) do
    {:ok, contents} -> Jason.decode!(contents)
    {:error, _} -> %{}  
  end

reading_fixes =
  case File.read(reading_fixes_file) do
    {:ok, contents} -> Jason.decode!(contents)
    {:error, _} -> %{}  
  end

all = Repo.all(Kanji)

string_strokes = Enum.filter(all, fn k ->
  strokes = get_in(k.stroke_data, ["strokes"])
  is_list(strokes) and length(strokes) > 0 and is_binary(hd(strokes))
end)

unknown_strokes = Enum.filter(all, fn k ->
  strokes = get_in(k.stroke_data, ["strokes"])
  is_list(strokes) and is_map(hd(strokes)) and (hd(strokes)["type"] == "unknown" or hd(strokes)["direction"] == "unknown")
end)

no_strokes = Enum.filter(all, fn k ->
  is_nil(k.stroke_data) or k.stroke_data == %{} or is_nil(k.stroke_data["strokes"]) or k.stroke_data["strokes"] == []
end)

no_readings = from(k in Kanji,
  left_join: r in assoc(k, :kanji_readings),
  where: not is_nil(k.stroke_data) and k.stroke_data != ^%{},
  group_by: k.id,
  having: count(r.id) == 0,
  select: k
) |> Repo.all()

chars_strokes = (string_strokes ++ unknown_strokes ++ no_strokes) |> Enum.map(& &1.character) |> Enum.uniq()
chars_readings = no_readings |> Enum.map(& &1.character) |> Enum.uniq()

IO.puts("Kanji needing strokes: #{length(chars_strokes)}")
IO.puts("Kanji needing readings: #{length(chars_readings)}")

# Build module lines
lines = []
lines = lines ++ ['defmodule Medoru.Content.KanjiBulkFix do']
lines = lines ++ ['  @moduledoc """']
lines = lines ++ ['  Self-contained bulk fix for kanji strokes and readings.']
lines = lines ++ ['  Run: Medoru.Content.KanjiBulkFix.apply()']
lines = lines ++ ['  """']
lines = lines ++ ['']
lines = lines ++ ['  alias Medoru.Repo']
lines = lines ++ ['  alias Medoru.Content.{Kanji, KanjiReading}']
lines = lines ++ ['  import Ecto.Query']
lines = lines ++ ['  require Logger']
lines = lines ++ ['']

# Stroke fixes
lines = lines ++ ['  @stroke_fixes %{']
for char <- chars_strokes do
  strokes = Map.get(stroke_fixes, char)
  if strokes && strokes["strokes"] && length(strokes["strokes"]) > 0 do
    lines = lines ++ ['    "#{char}" => [']
    for s <- strokes["strokes"] do
      path = String.replace(s["path"], "\"", "\\\"")
      lines = lines ++ ['      %{path: "#{path}", order: #{s["order"]}, type: "#{s["type"]}", direction: "#{s["direction"]}"},']
    end
    lines = lines ++ ['    ],']
  end
end
lines = lines ++ ['  }']
lines = lines ++ ['']

# Reading fixes
lines = lines ++ ['  @reading_fixes %{']
for char <- chars_readings do
  readings = Map.get(reading_fixes, char)
  if readings && length(readings) > 0 do
    lines = lines ++ ['    "#{char}" => [']
    for r <- readings do
      lines = lines ++ ['      %{type: :#{r["type"]}, reading: "#{r["reading"]}", romaji: "#{r["romaji"]}"},']
    end
    lines = lines ++ ['    ],']
  end
end
lines = lines ++ ['  }']
lines = lines ++ ['']

# Functions
lines = lines ++ ['  def apply do']
lines = lines ++ ['    apply_stroke_fixes()']
lines = lines ++ ['    apply_reading_fixes()']
lines = lines ++ ['  end']
lines = lines ++ ['']
lines = lines ++ ['  defp apply_stroke_fixes do']
lines = lines ++ ['    Logger.info("Applying stroke fixes...")']
lines = lines ++ ['']
lines = lines ++ ['    string_stroke_kanji =']
lines = lines ++ ['      Repo.all(Kanji)']
lines = lines ++ ['      |> Enum.filter(fn k ->']
lines = lines ++ ['        strokes = get_in(k.stroke_data, ["strokes"])']
lines = lines ++ ['        is_list(strokes) and length(strokes) > 0 and is_binary(hd(strokes))']
lines = lines ++ ['      end)']
lines = lines ++ ['']
lines = lines ++ ['    {fixed_kv, fixed_fb} =']
lines = lines ++ ['      Enum.reduce(string_stroke_kanji, {0, 0}, fn kanji, {kv, fb} ->']
lines = lines ++ ['        case Map.fetch(@stroke_fixes, kanji.character) do']
lines = lines ++ ['          {:ok, strokes} ->']
lines = lines ++ ['            current = kanji.stroke_data || %{}']
lines = lines ++ ['            new_data = Map.merge(current, %{"strokes" => strokes, "bounds" => %{"width" => 109, "height" => 109, "viewBox" => "0 0 109 109"}})']
lines = lines ++ ['            kanji |> Ecto.Changeset.change(stroke_data: new_data) |> Repo.update!()']
lines = lines ++ ['            {kv + 1, fb}']
lines = lines ++ ['']
lines = lines ++ ['          :error ->']
lines = lines ++ ['            strokes = kanji.stroke_data["strokes"] || []']
lines = lines ++ ['            new_strokes = Enum.with_index(strokes, 1) |> Enum.map(fn {path, idx} ->']
lines = lines ++ ['              %{"path" => path, "order" => idx, "type" => "unknown", "direction" => "unknown"}']
lines = lines ++ ['            end)']
lines = lines ++ ['            current = kanji.stroke_data || %{}']
lines = lines ++ ['            kanji |> Ecto.Changeset.change(stroke_data: Map.put(current, "strokes", new_strokes)) |> Repo.update!()']
lines = lines ++ ['            {kv, fb + 1}']
lines = lines ++ ['        end']
lines = lines ++ ['      end)']
lines = lines ++ ['']
lines = lines ++ ['    unknown_stroke_kanji =']
lines = lines ++ ['      Repo.all(Kanji)']
lines = lines ++ ['      |> Enum.filter(fn k ->']
lines = lines ++ ['        strokes = get_in(k.stroke_data, ["strokes"])']
lines = lines ++ ['        is_list(strokes) and is_map(hd(strokes)) and (hd(strokes)["type"] == "unknown" or hd(strokes)["direction"] == "unknown")']
lines = lines ++ ['      end)']
lines = lines ++ ['      |> Enum.filter(fn k -> Map.has_key?(@stroke_fixes, k.character) end)']
lines = lines ++ ['']
lines = lines ++ ['    replaced = Enum.reduce(unknown_stroke_kanji, 0, fn kanji, count ->']
lines = lines ++ ['      strokes = @stroke_fixes[kanji.character]']
lines = lines ++ ['      current = kanji.stroke_data || %{}']
lines = lines ++ ['      new_data = Map.merge(current, %{"strokes" => strokes, "bounds" => %{"width" => 109, "height" => 109, "viewBox" => "0 0 109 109"}})']
lines = lines ++ ['      kanji |> Ecto.Changeset.change(stroke_data: new_data) |> Repo.update!()']
lines = lines ++ ['      count + 1']
lines = lines ++ ['    end)']
lines = lines ++ ['']
lines = lines ++ ['    no_stroke_kanji =']
lines = lines ++ ['      Repo.all(Kanji)']
lines = lines ++ ['      |> Enum.filter(fn k ->']
lines = lines ++ ['        is_nil(k.stroke_data) or k.stroke_data == %{} or is_nil(k.stroke_data["strokes"]) or k.stroke_data["strokes"] == []']
lines = lines ++ ['      end)']
lines = lines ++ ['      |> Enum.filter(fn k -> Map.has_key?(@stroke_fixes, k.character) end)']
lines = lines ++ ['']
lines = lines ++ ['    added = Enum.reduce(no_stroke_kanji, 0, fn kanji, count ->']
lines = lines ++ ['      strokes = @stroke_fixes[kanji.character]']
lines = lines ++ ['      new_data = %{"strokes" => strokes, "bounds" => %{"width" => 109, "height" => 109, "viewBox" => "0 0 109 109"}}']
lines = lines ++ ['      kanji |> Ecto.Changeset.change(stroke_data: new_data) |> Repo.update!()']
lines = lines ++ ['      count + 1']
lines = lines ++ ['    end)']
lines = lines ++ ['']
lines = lines ++ ['    Logger.info("Strokes: #{fixed_kv} KanjiVG, #{fixed_fb} fallback, #{replaced} replaced, #{added} added")']
lines = lines ++ ['  end']
lines = lines ++ ['']
lines = lines ++ ['  defp apply_reading_fixes do']
lines = lines ++ ['    Logger.info("Applying reading fixes...")']
lines = lines ++ ['']
lines = lines ++ ['    kanji_without_readings =']
lines = lines ++ ['      from(k in Kanji,']
lines = lines ++ ['        left_join: r in assoc(k, :kanji_readings),']
lines = lines ++ ['        where: not is_nil(k.stroke_data) and k.stroke_data != ^%{},']
lines = lines ++ ['        group_by: k.id,']
lines = lines ++ ['        having: count(r.id) == 0,']
lines = lines ++ ['        select: k']
lines = lines ++ ['      )']
lines = lines ++ ['      |> Repo.all()']
lines = lines ++ ['']
lines = lines ++ ['    inserted =']
lines = lines ++ ['      Enum.reduce(kanji_without_readings, 0, fn kanji, count ->']
lines = lines ++ ['        case Map.fetch(@reading_fixes, kanji.character) do']
lines = lines ++ ['          {:ok, readings} when is_list(readings) and readings != [] ->']
lines = lines ++ ['            n = Enum.reduce(readings, 0, fn r, inner ->']
lines = lines ++ ['              attrs = %{kanji_id: kanji.id, reading_type: r.type, reading: r.reading, romaji: r.romaji}']
lines = lines ++ ['              case %KanjiReading{} |> KanjiReading.changeset(attrs) |> Repo.insert() do']
lines = lines ++ ['                {:ok, _} -> inner + 1']
lines = lines ++ ['                {:error, _} -> inner']
lines = lines ++ ['              end']
lines = lines ++ ['            end)']
lines = lines ++ ['            count + n']
lines = lines ++ ['']
lines = lines ++ ['          _ -> count']
lines = lines ++ ['        end']
lines = lines ++ ['      end)']
lines = lines ++ ['']
lines = lines ++ ['    Logger.info("Inserted #{inserted} readings")']
lines = lines ++ ['  end']
lines = lines ++ ['end']

output = Enum.join(lines, "\n") <> "\n"
out_path = "lib/medoru/content/kanji_bulk_fix.ex"
File.write!(out_path, output)

size_kb = byte_size(output) / 1024
IO.puts("\nWritten: #{out_path}")
IO.puts("Size: #{:erlang.float_to_binary(size_kb, decimals: 1)} KB")
IO.puts("Kanji with stroke fixes embedded: #{length(chars_strokes)}")
IO.puts("Kanji with reading fixes embedded: #{length(chars_readings)}")
